content/blog/2022-03-24-e2e-computer-vision.md

```diff
@@ -0,0 +1,420 @@
+   bills
+
+then the answer is yes if you use [CML library](https://cml.dev/) in combination
+with CI/CD tools like [GitHub Actions](https://docs.github.com/en/actions) or
+[GitLab CI/CD](https://docs.gitlab.com/ee/ci/).
+
+We've configured three
+[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/master/.github/workflows)
+for GitHub Actions, each of which corresponds to a particular stage depending on
+the project's lifecycle we are in:
+
+1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/train-model.yaml)
+
+   In this stage, we'll be working on an experimentation git branch:
+   experimenting with data preprocessing, changing model architecture, tuning
+   hyperparameters, etc. Once we think our experiment is ready to be run, we'll
+   push our changes to a remote repository (in this case, GitHub). This push
+   will trigger a CI/CD job in GitHub Actions, which in turn will:
+
+   1. provision an EC2 virtual machine with a GPU in AWS
+   2. deploy our experiment branch to this machine
+   3. rerun the entire DVC pipeline
+   4. push files with metrics and other DVC artifacts back to GitHub
+
+   At this point, we can assess the results in DVC Studio and GitHub and decide
+   what things we want to change next.
+
+```mermaid
+flowchart TB
+A(Commit to exp branch) -->|Push changes| B("Exp CML workflow\n(training & reporting)")
+B --> |Reports,\\nmetrics,\\nplots| C("Check results.\nAre they good?")
+C --> |No\\nchange experiment parameters | A
+C -->|Yes| D(PR to dev branch)
+```
+
+1. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/dev-train-upload-deploy.yaml)
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the dev branch. This would trigger a different CI/CD job
+   that will:
+
+   1. retrain the model with the new parameters
+   2. deploy the web REST API application (that relies on the new retrained
+      model) to a development endpoint on Heroku
+
+   Now we can test our API and assess the end-to-end performance of the overall
+   solution.
+
+```mermaid
+flowchart TB
+A(Merge to dev branch) --> B(Retraining) --> C(Deployment to dev environment)
+
+```
+
+1. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/prod-deploy-api-to-heroku.yaml)
+
+   If we've thoroughly tested and monitored our development web API, we can
+   merge the development branch into the main branch of our repository. Again,
+   this triggers the 3rd CI/CD workflow that deploys the code from the main
+   branch to the production API.
+
+```mermaid
+flowchart TB
+B(Merge dev into main) --> C(Prod CML workflow) --> D(Deployment to prod)
+
+```
+
+### Experiment Management with DVCLive and Studio
```

Comment:
Again, It feels like the above section could be expanded into a separate blog post. 

I would like to see the actual code (or a simplified version) corresponding to each of the bullet points that describe what each workflow does.